### PR TITLE
Remove unnecessary work from lazy loops (and a little from greedy)

### LINKED ISSUE #68415: Improve handling of lazy loops in regex source generator / compiler
Lazy loops currently push/pop more state on to the backtracking stack than is generally necessary.  This can be improved.

https://github.com/dotnet/runtime/blob/6aff1a3851a89f20dbc3f11e8724df2a4f059171/src/libraries/System.Text.RegularExpressions/gen/RegexGenerator.Emitter.cs#L3188-L3197

### PR DESCRIPTION
**Pull request by stephentoub**
- For a lazy loop, compute whether the child might match empty.  Only if it might do we need any of the checks around empty iterations, and only in those situations do we need to track the starting position or whether we've seen an empty loop, so avoid all work associated with that if the min length of the child is greater than zero.
- If the lazy loop failing will exit the match, avoid doing all the cleanup work that ends up being irrelevant.
- If the lazy loop isn't itself part of another loop, we don't need the backtracking section to push/pop any additional state on the backtracking stack.
- Avoid outputting two gotos in a row in greedy loops.
- Avoid outputting additional if blocks in greedy loops when the checks are identical.
- Add some comments, in particular to loops where the logic can be tricky and hard to follow / reason about.
- Remove some emitted semicolons after labels.
- Fix the naming of a few "SkipBacktrack" labels to conform to the same prefixed naming convention used for other labels

Fixes https://github.com/dotnet/runtime/issues/68415

### CODE REVIEW

File: src/libraries/System.Text.RegularExpressions/gen/RegexGenerator.Emitter.cs [hunk 1]
```diff
@@ -3199,13 +3205,25 @@ void EmitLazy(RegexNode node)
...
+                string iterationCount = ReserveName("lazyloop_iteration");
+                additionalDeclarations.Add($"int {iterationCount} = 0;");
+                writer.WriteLine($"{iterationCount} = 0;");
+
+                // Loops that match empty iterations need additional checks in place to prevent infinitely matching (since
+                // you could end up looping an infinite number of times at the same location).  We can avoid those
+                // additional checks if we can prove that the loop can never match empty, which we can do by computing
+                // the minimum length of the child; only if it's 0 might iterations be empty.
+                bool iterationMayBeEmpty = child.ComputeMinLength() == 0;
+                string startingPos = null, sawEmpty = null;
```
Comment (joperezr) [inline src/libraries/System.Text.RegularExpressions/gen/RegexGenerator.Emitter.cs:3220]: `string?`
Is the generator not using nullable annotations? If not I suppose we should enable them right?

File: src/libraries/System.Text.RegularExpressions/gen/RegexGenerator.Emitter.cs [hunk 2]
```diff
@@ -3199,13 +3205,25 @@ void EmitLazy(RegexNode node)
...
+                if (iterationMayBeEmpty)
+                {
+                    startingPos = ReserveName("lazyloop_starting_pos");
+                    sawEmpty = ReserveName("lazyloop_empty_seen");
```
Comment (joperezr) [inline src/libraries/System.Text.RegularExpressions/gen/RegexGenerator.Emitter.cs:3224]: NIT: the variable name is a bit confusing since it is a `int` as opposed to a `bool`. Should we use something like `laoop_empty_seen_count` or something like that?

Comment (stephentoub) [reply src/libraries/System.Text.RegularExpressions/gen/RegexGenerator.Emitter.cs:3224]: It is a Boolean, in that it can only be 0 or 1.  It's just stored as an int because that's what our stack stores.

File: src/libraries/System.Text.RegularExpressions/gen/RegexGenerator.Emitter.cs (hunk 1 above)
Comment (stephentoub) [reply src/libraries/System.Text.RegularExpressions/gen/RegexGenerator.Emitter.cs:3220]: > Is the generator not using nullable annotations?

It has annotations enabled but warnings disabled.

> If not I suppose we should enable them right?

It targets netstandard2.0, where things like Debug.Assert aren't annotated with [DoesNotReturnWhen(false)], so the number of false positive warnings is staggering.

At one point I temporarily retargeted to .NET 7 and fixed the nullable warnings.  We can do another pass like that.

**Review by joperezr — APPROVED** [review]
LGTM. I assume we already have more than enough coverage around all of these branches with our existing testbed.
